### DISCUSSION TITLE
When to upgrade servers?

### DISCUSSION THREAD

**Discussion by mumbel** [NationalSecurityAgency/ghidra#6642] [General]
I've never seen any compatibility issues on my server.  We ran 9.0.1 or 9.1 (i forget) for the longest time and upgraded when 10.0.1 came out.  Mostly just due to the aging VM we originally installed on (maybe this also coincided with 18.04 EOL).

Should users be upgrading their servers more often?  Is there any performance increase by upgrading due to changes in code, underlying JDK, or other packages (i.e. yajsw... ps looks like a few versions have come out since ghidra's current) or really not enough going on in the server to impact clients?

Just curious, since this last release mentioned updates to performance of large amounts of files in the shared repo.  Is it beneficial to just upgrade the client or would the setup be improved with a server upgrade as well?

**Comment by ryanmkurtz** [comment]
Will ping @ghidra1 for an expert answer.

**Comment by ghidra1** [comment]
We strive to retain compatibility of the server interface and it rarely changes thus upgrades are rarely required.  Server upgrades can impact a large number of user's so any such upgrade is left to the discretion of the user's project team.  The What's New document will indicate compatibility for the current release or if a server upgrdae is recommended.  

Aside from the server interface, the major version of Java used by the Ghidra Server should match that used by all the Ghidra Clients.

**Comment by mumbel** [comment]
This is a non answer. I asked specifically about performance and improvements and didn't not ask about compatibility, I even acknowledged I had never had compatibility issues by using the same version for so long.

So ghidra requiring 11 is not compatible with ghidra requiring 17 and will not be compatible with upcoming release using 21?

**Reply by ghidra1** [reply]
> So ghidra requiring 11 is not compatible with ghidra requiring 17 and will not be compatible with upcoming release using 21?

My statement regarding Java version consistency between client and server was incorrect.  I believe the Java versions used between client and server can differ.  Although, there may be issues related to TLS compatibility that have occurred in the past which we had attributed to the Java version in use and the required TLS version.

Our locked-in serialVersionUID for those classes which get serialized should prevent java build version from being an issue.

**Comment by ghidra1** [comment]
The performance improvements which were made in Ghidra 11.1 were client-side only and do not require a server upgrade to benefit from them.  There have never been any server-side performance improvements.
